### PR TITLE
[RHCLOUD-35383] Add v2_role FK, indexes, and unique constraint to BindingMapping

### PR DESCRIPTION
**Pull request by dsova06**
## Link(s) to Jira
- [RHCLOUD-35383](https://redhat.atlassian.net/browse/RHCLOUD-35383)

## Description of Intent of Change(s)

**WHAT:** Add a ForeignKey from BindingMapping to RoleV2, compound indexes, and a unique constraint to prevent duplicate binding mappings.

**WHY:** BindingMapping currently stores the V2 role ID buried inside a JSON field (`mappings["role"]["id"]`) and has zero indexes or constraints. This causes inefficient lookups, no database-level prevention of duplicate bindings, and required a complex locking workaround in the dual-write handler to prevent concurrent duplicate creation.

**HOW:**
- Add nullable `v2_role` ForeignKey to BindingMapping pointing to RoleV2 (SET_NULL on delete)
- Add compound index on `(role, resource_type_namespace, resource_type_name, resource_id)` for V1 role queries used in the locking code and workspace service
- Add unique constraint on `(v2_role, resource_type_namespace, resource_type_name, resource_id)` to prevent duplicate bindings per V2 role and resource
- Data migration backfills `v2_role` from `mappings["role"]["id"]` by looking up RoleV2 records by UUID
- Update `for_role_binding()` factory method to accept optional `v2_role` parameter
- Update `sharedSystemRolesReplicatedRoleBindings` caller to pass `v2_role`

Three migrations ordered for safety: schema (add field + index) -> data (backfill) -> constraint (unique). PostgreSQL treats NULL as distinct in unique constraints, so rows without a v2_role won't conflict.

SET_NULL is used instead of CASCADE to preserve BindingMappings when a RoleV2 is deleted. The V1 `role` FK already handles cleanup via CASCADE when a V1 role is deleted. SET_NULL allows lazy V2 role re-creation from existing mappings (used by the dual-write handler's custom role migration path).

**Out of scope (follow-up):**
- Removing the resource-level locking workaround in `relation_api_dual_write_subject_handler.py` (behavioral change, separate PR)
- Replacing `mappings["role"]["id"]` JSON lookups with `v2_role` FK queries
- Setting `v2_role` in `_create_default_mapping_for_system_role`

## Local Testing
- `pipenv run tox -e lint` -- passes
- `pipenv run tox -e py312-fast -- tests.management.role.test_model.BindingMappingTests` -- all 20 tests pass (6 new)
- `pipenv run tox -e py312-fast` -- full suite passes
- Migrations applied and verified against local PostgreSQL: FK column, compound index, and unique constraint all present

## Checklist
- [ ] if API spec changes are required, is the spec updated? N/A
- [ ] are there any pre/post merge actions required? No
- [x] are these changes covered by unit tests?
- [ ] if warranted, are documentation changes accounted for? N/A
- [x] does this require migration changes?
  - [x] if yes, are they backwards compatible? Yes -- new nullable FK, additive index and constraint
- [ ] is there known, direct impact to dependent teams/components? No

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

[RHCLOUD-35383]: https://redhat.atlassian.net/browse/RHCLOUD-35383?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by dsova06 — COMMENTED** [review]
The migration ordering (schema -> data -> constraint) is solid, and the test coverage hits the important edge cases -- NULL semantics, cascade behavior, factory backward compatibility. Clean scope with follow-ups explicitly called out.

### Would be good to clean up

1. **Data migration backfill does individual saves** -- The backfill in `0089` does `bm.save(update_fields=["v2_role_id"])` per row inside an `iterator(chunk_size=1000)` loop. The chunking helps on the read side but writes still go one at a time. If there are many BindingMapping rows in production, this could slow down the deployment window. Batching with `bulk_update` would cut the DB round-trips significantly:

```python
batch = []
for bm in BindingMapping.objects.filter(v2_role__isnull=True).iterator(chunk_size=1000):
    # ... lookup logic ...
    bm.v2_role_id = v2_role_pk
    batch.append(bm)
    if len(batch) >= 1000:
        BindingMapping.objects.bulk_update(batch, ["v2_role_id"])
        updated += len(batch)
        batch = []
if batch:
    BindingMapping.objects.bulk_update(batch, ["v2_role_id"])
    updated += len(batch)
```

### Something to think about

1. **Type hint on `v2_role` parameter** -- `for_role_binding` has type annotations for the other parameters but `v2_role=None` is untyped. An `Optional["RoleV2"] = None` would keep it consistent with the rest of the method signature.

2. **Dual-write handler still creates NULL v2_role rows** -- `_create_default_mapping_for_system_role` in `relation_api_dual_write_subject_handler.py` calls `for_role_binding(binding, system_role)` without passing `v2_role`. This is explicitly out of scope per the PR description, but worth tracking since the dual-write path is the main runtime creator of new BindingMappings -- without a follow-up, new rows will keep accumulating with NULL v2_role after the backfill runs.

Looks good otherwise. Nice clean separation of the three migrations.

**Comment by dsova06** [comment]
Addressing Sourcery's feedback:

1. **`v2_role` not saved in `_get_or_create_binding`**: This is by design. `update_mappings_from_role_binding` also doesn't persist -- it just sets `self.mappings` in memory. Both branches of `_get_or_create_binding` return unsaved instances. The caller (`permission_groupings_to_v2_role_bindings`) collects them into `MigrateCustomRoleResult` and the ultimate consumer saves. Same pattern as the new-mapping branch which creates via `for_role_binding()` without saving.

2. **In-memory dict of RoleV2**: Fixed. The backfill now collects needed UUIDs from BindingMapping first, then queries only those RoleV2 records.

3. **Constraint name with spaces**: Fixed. Renamed to `unique_bindingmapping_v2role_resource`.

Also addressed self-review feedback:
- Data migration now uses `bulk_update` with batching instead of individual `save()` calls
- Added `Optional["RoleV2"]` type hint to `for_role_binding`'s `v2_role` parameter

**Comment by dsova06** [comment]
@sourcery-ai resolve

**Comment by randomnetcat** [comment]
Keeping two redundant ways of storing the V2 role around unnerves me a bit. I think the way to do this while keeping everything normalized would be to add a computed column for the role UUID in `BindingMapping`, then have the foreign be from the computed column onto the `RoleV2` UUID.

Before doing this, we'd also need to remove old BindingMappings for V2 tenants (and delete any new ones on conversion), since they could reference since-deleted V2 roles.

**Comment by dsova06** [comment]
Good point about the redundancy @randomnetcat. I looked into the computed column approach -- PostgreSQL does support FKs on generated columns, but Django's ORM can't express it (GeneratedField can't be used as a ForeignKey). We'd need raw SQL migrations and a model that pretends the column is a regular field, which adds maintenance burden for anyone touching migrations later.

The way I see the tradeoff: the redundancy is temporary. There are only ~2 files outside the model that read `mappings["role"]["id"]` (`internal/utils.py` and the migration tool). The plan is to replace those lookups with `v2_role` FK queries in follow-up PRs, at which point the JSON role data becomes dead weight we can stop writing.

Going the computed column route solves the sync problem, but that benefit disappears once we stop writing the role into JSON. We'd be fighting the ORM to avoid a transitional state that gets cleaned up anyway.

On the orphaned references -- the data migration already handles this by skipping rows where the RoleV2 doesn't exist (they keep `v2_role=NULL`). The FK is nullable with SET_NULL, so no constraint violations.

Happy to track the follow-up cleanup as a separate ticket if that helps make the path clearer.

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            skipped += 1
+            continue
+
+        bm.v2_role_id = v2_role_pk
+        batch.append(bm)
+
+        if len(batch) >= 1000:
+            BindingMapping.objects.bulk_update(batch, ["v2_role_id"])
+            updated += len(batch)
+            batch = []
```
Comment (randomnetcat) [inline rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: `itertools.batched` might be a cleaner way to express this?

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        if role_data and "id" in role_data:
+            needed_uuids.add(str(role_data["id"]))
+
+    v2_role_lookup = {str(r.uuid): r.pk for r in RoleV2.objects.filter(uuid__in=needed_uuids).only("id", "uuid")}
+
+    batch = []
+    updated = 0
+    skipped = 0
+
+    for bm in BindingMapping.objects.filter(v2_role__isnull=True).iterator(chunk_size=1000):
```
Comment (randomnetcat) [inline rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: We need to lock the `BindingMapping`s `FOR UPDATE` here.

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+def backfill_v2_role(apps, schema_editor):
+    """Backfill v2_role FK from mappings JSON field."""
+    BindingMapping = apps.get_model("management", "BindingMapping")
+    RoleV2 = apps.get_model("management", "RoleV2")
+
+    needed_uuids = set()
+    for bm in BindingMapping.objects.filter(v2_role__isnull=True).iterator(chunk_size=1000):
+        role_data = bm.mappings.get("role")
+        if role_data and "id" in role_data:
+            needed_uuids.add(str(role_data["id"]))
```
Comment (randomnetcat) [inline rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: `BindingMapping.objects.values_list("mappings__role__id", flat=True).distinct()`?

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    for bm in BindingMapping.objects.filter(v2_role__isnull=True).iterator(chunk_size=1000):
+        role_data = bm.mappings.get("role")
+        if not role_data or "id" not in role_data:
+            skipped += 1
+            continue
+
+        v2_role_pk = v2_role_lookup.get(str(role_data["id"]))
+        if v2_role_pk is None:
+            skipped += 1
+            continue
```
Comment (randomnetcat) [inline rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: There are race conditions here:
* We could find a `BindingMapping` created after we pre-loaded all of the roles (in which case we'd need to fall back to looking it up).
* Currently, we could find a `BindingMapping` that references a since-deleted role, but I believe that would be fixed by locking.

File: rbac/management/role/model.py [hunk 5]
```diff
@@ -153,10 +156,31 @@ class BindingMapping(models.Model):
     # JSON encoding of migration_tool.models.V2rolebinding
     mappings = models.JSONField(default=dict)
     role = models.ForeignKey(Role, on_delete=models.CASCADE, related_name="binding_mappings")
+    v2_role = models.ForeignKey(
+        "management.RoleV2",
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="binding_mappings",
+    )
```
Comment (randomnetcat) [inline rbac/management/role/model.py:165]: So, logically, the correct `on_delete` value here would be `PROTECT`. As I alluded to in my initial comment, doing this would require us to fully tear down `BindingMapping`s when a tenant migrates to V2. We could proceed like this for now, but we should work on getting it updated in the near future.

File: rbac/management/role/model.py [hunk 6]
```diff
@@ -167,7 +191,9 @@ def save(self, *args, **kwargs):
         super().save(*args, **kwargs)
 
     @classmethod
-    def for_role_binding(cls, role_binding: V2rolebinding, v1_role: Union[Role, str]):
+    def for_role_binding(
+        cls, role_binding: V2rolebinding, v1_role: Union[Role, str], v2_role: Optional["RoleV2"] = None
```
Comment (randomnetcat) [inline rbac/management/role/model.py]: `v2_role` seems like it shouldn't be optional?

File: rbac/management/migrations/0090_bindingmapping_unique_v2_role_resource.py
Comment (randomnetcat) [inline rbac/management/migrations/0090_bindingmapping_unique_v2_role_resource.py:1]: This doesn't need to be a separate migration. The unique constraint won't apply to rows that contain a null in one of the relevant columns.

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py (hunk 1 above)
Comment (dsova06) [reply rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: Fixed: switched to `itertools.batched` from stdlib (Python 3.12+). Removes manual batch tracking entirely.

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py (hunk 2 above)
Comment (dsova06) [reply rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: Fixed: added `select_for_update()` on the update queryset to prevent concurrent modification during backfill.

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py (hunk 3 above)
Comment (dsova06) [reply rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: Fixed: replaced the Python loop with `values_list("mappings__role__id", flat=True)` to collect UUIDs in a single query.

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py (hunk 4 above)
Comment (dsova06) [reply rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: Fixed: added a fallback individual lookup when a UUID is missing from the pre-built lookup dict. This handles rows that appeared after the initial UUID collection pass. The result is cached in the lookup dict so repeated misses for the same UUID don't cause N+1 queries.

File: rbac/management/role/model.py (hunk 5 above)
Comment (dsova06) [reply rbac/management/role/model.py:165]: Agreed -- PROTECT is the right long-term target. For now SET_NULL is needed because the dual-write handler's lazy migration path (relation_api_dual_write_subject_handler.py:414) deletes and re-creates CustomRoleV2 records, and PROTECT would block that deletion. Once we remove the lazy migration path (after full backfill), switching to PROTECT becomes safe. Tracking as a follow-up.

File: rbac/management/role/model.py (hunk 6 above)
Comment (dsova06) [reply rbac/management/role/model.py]: Agreed on the direction -- v2_role should eventually be non-nullable. It depends on the backfill completing across all tenants and the lazy migration path being removed. Once those are done, we can add a migration to enforce NOT NULL. Keeping it optional for now to allow incremental rollout.

File: rbac/management/migrations/0090_bindingmapping_unique_v2_role_resource.py
Comment (dsova06) [reply rbac/management/migrations/0090_bindingmapping_unique_v2_role_resource.py:1]: Fixed: merged the unique constraint from migration 0090 into 0088 (the schema migration). PostgreSQL excludes NULLs from unique constraints, so this is safe to apply before the backfill. Deleted migration 0090.

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+    needed_uuids = set(
+        BindingMapping.objects.filter(v2_role__isnull=True)
+        .exclude(mappings__role__id=None)
+        .values_list("mappings__role__id", flat=True)
+    )
+
+    v2_role_lookup = {str(r.uuid): r.pk for r in RoleV2.objects.filter(uuid__in=needed_uuids).only("id", "uuid")}
+
+    rows = BindingMapping.objects.filter(v2_role__isnull=True).select_for_update()
```
Comment (randomnetcat) [inline rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: First, does this run in a transaction? `select_for_update` is meaningless if not.

Second, this will eventually result in us holding locks on *all* of the BindingMappings; that isn't a good idea. See, e.g., `_remove_orphaned_role_bindings` for a way to avoid this (do a queryset for all of the entries, then lock and release them in batches).

File: rbac/management/role/model.py (hunk 6 above)
Comment (randomnetcat) [reply rbac/management/role/model.py]: The *field* being optional doesn't mean this parameter has to be. Do we not want to enforce that it's being set in new `BindingMapping`s from here on out?

File: rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py (hunk 7 above)
Comment (dsova06) [reply rbac/management/migrations/0089_backfill_bindingmapping_v2_role.py]: Fixed: restructured to batch lock-and-release. Collect all unprocessed PKs upfront, then iterate in chunks of 1000 with `transaction.atomic()` + `select_for_update()` inside each chunk. Locks are released at the end of each batch's transaction, following the same pattern as `_remove_incorrect_workspace_parent_relations`.

File: rbac/management/role/model.py (hunk 6 above)
Comment (dsova06) [reply rbac/management/role/model.py]: Fixed: made `v2_role` a required parameter (no default). Callers must now explicitly pass it -- even if they pass `None`. Updated the dual-write handler to look up the `SeededRoleV2` by UUID for system roles, and updated all test call sites to pass `v2_role=None` explicitly.
